man/pututmpid.8
.TH pututmpid 8 "Dec 28, 2009" 
.SH NAME
pututmpid \- write utmp/wtmp records and exec command
.SH SYNOPSIS
.B putupmpid
.I [-w]
.I ut_id
.I command
.I [parameters]
.br
.B pututmpid
.I reboot
.br
.B pututmpid
.I halt


.SH DESCRIPTION
.B pututmpid
is used to write records in utmp/wtmp files for running 
getty/agetty/fgetty.
pututmpid writes the record and then execve()s a command.
The commands 
.B login 
and 
.B getty
use the records written from pututmpid.

After reboot it is possible to set 
.B pututmpid reboot
in init scripts.
This will update /var/run/utmp with BOOT_TIME record.  This
must be done after /var/run/utmp is writable.  The usage of
.B pututmpid halt
is similar.  It sets wtmp record (in /var/log/wtmp file).

.SH OPTIONS
usually pututmpid is symlinked as \fIrun\fR command of a service.
.TP
.B \-w
debug mode -- write more info in /var/log/wtmp
.TP
.B ut_id
string at most 4 bytes long.  examples: tty1,tty2, vc1,vc2,vc3...
Use different names for each virtual console!
.TP
.B command
the real command to start
.TP
.B parameters
additional parameters for the command

.SH EXAMPLE
A typical use of this command will look like this:
.TP
/etc/ninit/getty/2
.TP
/etc/ninit/getty/2/params
tty2
.br
/sbin/fgetty
.br
tty2
.TP
/etc/ninit/getty/2/run
-> /sbin/pututmpid

.SH "SEE ALSO"
ninit(8), ninit\-runlevel(8)
.br
getty(8), {n|f|a|min}getty(8),
.br
login(1), utmp(5), runlevel(8)


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


pututmpid.c
/* pututmpid.c

   usage: pututmpid [-w] c2 child arg... 
   usage: pututmpid reboot
   usage: pututmpid halt

   c2 is the same entry as in /etc/inittab
   c2:2345:respawn:/sbin/fgetty tty2
   see the end of this file how it works
*/

#include <unistd.h>
#include <fcntl.h>
#include <utmp.h>
#include <time.h>
#include <sys/utsname.h>
#include "ninitfeatures.h"
#include "utmp_defs.h"

int main(int argc, char **argv) {
  struct utmp u;
  char *utid,*argv0,*x;
  char flagwtmp=0;
  int fd; 
  off_t pos=0;

  if (argv[1] && argv[1][0] == '-' && argv[1][1] == 'w') 
    {++argv; --argc; flagwtmp=1;}

  if (argc >=3) {
    char **e, *env[2] = {0,0};
    size_t len = str_len(argv[1]);
    utid = argv[1];
    if (len>4) utid += len-4; 
    argv += 2;

    fd=open(_PATH_UTMP, O_RDWR);
    while (utmp_io(fd, &u, F_RDLCK)) {
      if (!str_diffn(u.ut_id, utid, sizeof(u.ut_id))) {
	if (u.ut_type == USER_PROCESS) {
	  u.ut_tv.tv_sec=time(0);
	  u.ut_type=DEAD_PROCESS;
	  /* byte_zero(&u.ut_user, sizeof(u.ut_user));
	     byte_zero(&u.ut_host, sizeof(u.ut_host)); */
	  do_wtmp(&u);
	}
	break ; 
      }
      pos += sizeof(struct utmp);
    }
    
    byte_zero(&u,sizeof(u));
    str_copyn(u.ut_id, utid, sizeof(u.ut_id)); 
    u.ut_pid=getpid();
    u.ut_tv.tv_sec=time(0);
    u.ut_type=INIT_PROCESS;
    
    if (fd>=0 && lseek(fd,pos,SEEK_SET) == pos)
      utmp_io(fd,&u,F_WRLCK);
    close(fd);
    
    if (flagwtmp) do_wtmp(&u);
    
    argv0 = argv[0];
    x = argv[0];
    x += str_rchr(x,'/');
    if (x[0]) argv[0] = x+1;
    else argv[0] = argv0;

    for (e=environ; *e; e++) 
      if (!str_diffn(*e,"TERM=",5)) {*env=*e; break;}

    execve(argv0, argv, env);
    write(2,"unable to exec: ",16);
    write(2,argv0,str_len(argv0));
    write(2,"\n",1);
    return(100);
  }

  /******************************************************/
  if (argc == 2) {
    struct utsname uname_buf;
    fd = -1;
    switch (argv[1][0]) {
    case 'r':
      fd=open(_PATH_UTMP, O_RDWR);
      while (utmp_io(fd, &u, F_RDLCK)) {
	/* if already exist do nothing */
	if (u.ut_type == BOOT_TIME) { close(fd); return 111; }
	pos += sizeof(struct utmp);
      }
      
      byte_zero(&u,sizeof(u));
      str_copy(u.ut_user,"reboot");
      str_copy(u.ut_line,"~");
      u.ut_type=BOOT_TIME;
      break;
    case 's':
    case 'h':
      str_copy(u.ut_user,"shutdown");
      u.ut_type=RUN_LVL;
      str_copy(u.ut_line,"~~");
      break;
    default:
      goto usage;
    }
    
    str_copy(u.ut_id,"~~");
    u.ut_pid=0;
    u.ut_tv.tv_sec=time(0);
    
    /* Put the OS version in place of the hostname */
    if (uname(&uname_buf) == 0)
      str_copyn(u.ut_host, uname_buf.release, 32); /* XXX overflow ? - 
                                                      very slim in Linux, 
                                                      see uname(2) and utmp(5).
                                                      However, this possibility exists
                                                      when the uname field length is
                                                      longer than the host name (not so in Linux) */
 
    if (fd>=0 && lseek(fd,pos,SEEK_SET) == pos)
      utmp_io(fd,&u,F_WRLCK);
    close(fd);
    
    do_wtmp(&u);
    return(0);
  }

 usage:
  write(2,"usage: pututmpid [-w] ut_id child arg...\n\
       pututmpid reboot\n\
       pututmpid halt\n",87);
  return(100);
}

/* 
    pututmpid:  
        if exist an entry in utmp with: ut_id=c2 ut_type=USER_PROCESS
          then put: ut_id=c2 ut_type=DEAD_PROCESS  in wtmp

        always put: ut_id=c2 ut_type=INIT_PROCESS  in utmp
    if -w flag put: ut_id=c2 ut_type=INIT_PROCESS  in wtmp 

    getty and login chanegs /var/run/utmp: 
    getty:          ut_id=c2 ut_type=LOGIN_PROCESS
    login:          ut_id=c2 ut_type=USER_PROCESS
*/
